Improve multi workspace handling

doudou opened the pull request:
This pull request improves having multiple Autoproj workspaces in a single VSCode workspace. In addition, it workarounds an issue related to VSCode's handling of the very first folder in the workspace.

First the vscode issue. VSCode will restart the extension hosts if the toplevel folder in a VSCode workspace changes (because it needs to change the now-obsolete `rootPath` property). This leads to the extension attempting to start a Watch while there is already one. The first attempt was to make sure we deregister the watch task using the new Task execution APIs. This is a great addition in the first place, but proved fruitless as it seems that vscode doesn't shut down the extensions cleanly (I wouldn't get the extension's `deactivate` function or any of the `dispose` methods to be called). I then tried to use the `taskExecutions` property to find the watch task, but it is also empty - don't really know why.

_In fine_, I tried to see a way to make sure the first folder doesn't change this much. This led me to think about a problem I recently had, to use two autoproj workspaces in the same vscode workspace (comparing code). This was hell, as the packages would be named the same.

So, I went for:
- always having the workspace's `autoproj` folder added to the vscode workspace, and name it as the workspace (e.g. `autoproj (squidbot)`).
- use the `autoproj` folder as a guard, i.e. folders for the squidbot workspace would be placed after the `autoproj (squidbot)` folder. Moreover, they would be placed **before** the following folder that is not part of the workspace project.

Because the folder names can't be changed after they have been added, I also created a `Rock: Add Workspace` command for the initial add. The one corner case is if someone adds the autoproj folder directly.

Illustration:

![vscode-multi-workspace](https://user-images.githubusercontent.com/292/53565196-86a96800-3b37-11e9-9e1d-8ab1c8eebcc8.png)

g-arjones commented:
> This leads to the extension attempting to start a Watch while there is already one

I handled this by:

1. [Keeping track of "Watch" tasks](https://github.com/g-arjones/vscode-autoproj/blob/0ce63254f50e6548f1cd7bf5434af529fb4dc4ba/src/extension.ts#L41)
2. [Killing watch tasks when the extension gets disposed](https://github.com/g-arjones/vscode-autoproj/blob/0ce63254f50e6548f1cd7bf5434af529fb4dc4ba/src/extension.ts#L29)

Which seems to work fine, AFAICT.

> Because the folder names can't be changed after they have been added

I don't think this is correct... You can rename the folder by doing:

```typescript
vscode.workspace.updateWorkspaceFolders(0, index, { name: 'foo', uri: folderUri });
```

[See this](https://github.com/Microsoft/vscode/blob/0a4a3b98282175a82a819b70d2bbe46475e1ebd0/src/vs/vscode.d.ts#L7431)

Still, I like the idea of having the main configuration automatically added when a new pacakage is added (and removed when all packages from the workspace get removed)

g-arjones commented on package.json:
```diff
@@ -45,6 +45,18 @@
+            "type": "object",
+            "title": "Rock Workspace Configuration",
+            "properties": {
+                "rock.devFolder": {
```
I don't understand what is this for

doudou replied:
It's just a simple way to be already in your main dev folder when doing "Add Workspace".

g-arjones replied:
Ah, cool..

doudou commented:
> (and removed when all packages from the workspace get removed)

Actually, that's something I really don't want to do.

doudou commented on src/commands.ts:
```diff
@@ -217,21 +274,24 @@ export class Commands
+        }
+
+        let folder = { name: name, uri: vscode.Uri.file(selectedOption.pkg.srcdir) };
+        let success = this._vscode.updateWorkspaceFolders(
```
Well ... since the extension is the one that provide the user with a list, it sounds a little bit too defensive, no ?

doudou replied:
Well ... since the extension is the one providing the user with a list, I'm not sure how this could happen.

g-arjones replied:
Yes, what I said makes no sense at all

doudou commented:
>     2\. Killing watch tasks when the extension gets disposed

Well, then, I don't know why, but according to the tests I did yesterday, the extension doesn't get disposed in the case where the first workspace folder is changed.

g-arjones commented:
> but through tasks instead of relying on the PID file

Not file, just a Map() from workspacePath to pid.

dispose() does get called but you can't do anything async in it (i.e use vscode's own terminateTask api) because apparently the extension host won't get the chance to execute that

doudou commented:
> dispose() does get called but you can't do anything async in it (i.e use vscode's own terminateTask api) because apparently the extension host won't get the chance to execute that

That would explain.

doudou commented:
> dispose() does get called but you can't do anything async in it (i.e use vscode's own terminateTask api) because apparently the extension host won't get the chance to execute that

`terminate()` is not even async ... they really are not letting the extensions do any form of cleanup.

doudou commented:
Anyways:
- updated the description for devFolder
- added a force-kill on dispose. I kept the "clean" terminate as well, it works in cases the workspace is disposed from within the normal workflow (i.e. last package removed from workspace)

g-arjones commented:
Looks good to me.. I liked the idea of creating one disposable per watch task.
Cool stuff... 👍

doudou commented:
>  I liked the idea of creating one disposable per watch task.

After working with vscode, I "stole" the disposable pattern in Roby itself. It's a really neat pattern.
